pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1264,30 +1263,85 @@ func (c *Cloud) getMountDevice(
 	// Check to see if this volume is already assigned a device on this machine
 	for mountDevice, mappingVolumeID := range deviceMappings {
 		if volumeID == mappingVolumeID {
-			if assign {
-				glog.Warningf("Got assignment call for already-assigned volume: %s@%s", mountDevice, mappingVolumeID)
-			}
 			return mountDevice, true, nil
 		}
 	}
+	return mountDevice(""), false, nil
+}
+
+// attachNewDevice attached a new disk to the node.
+func (c *Cloud) attachNewDevice(i *awsInstance, info *ec2.Instance, volumeID awsVolumeID) (mountDevice, error) {
+	// pick a new device name from pool of 52 devices (ba..bz,ca..cz). We are trying to prevent
+	// device name reuse as much as possible of recently freed devices. Reducing device name
+	// reuse on AWS helps prevent volume stuck in 'attaching' state bug.
+	chosen, deviceAllocator, alreadyAttached, err := c.allocateDiskDevice(i, info, volumeID)
+
+	if err != nil {
+		return "", err
+	}
+
+	glog.V(2).Infof("Assigned mount device %s -> volume %s", chosen, volumeID)
+	if !alreadyAttached {
+		attachResult, attachErr := c.performAwsAttach(chosen, volumeID, i)
+		if attachResult && deviceAllocator != nil {
+			// We only 'mark' device as taken when attach succeeds. We don't want failing attach
+			// calls to use device names.
+			deviceAllocator.MarkForUse(chosen)
```

Comment:
I am thinking, is Attach operation the right time to put a device to back of the queue? Should we de-prioritize a device when it was *detached* (or attach failed)?

When I have:

* One long running pod with /dev/xvdbb that starts at time 1:00 and finishes at 2:00.
* 51 short-running pods with all the other devices that started at time 1:01 and finished at time 1:02.

At time 2:01, all devices are free. Which one should I prefer when running a new pod? The one that was *attached* first or the one that was *detached* first? You could perhaps call MarkForUse both when attaching and when detaching.